### PR TITLE
Support grouping in drawio grouped endpoints elems

### PR DESCRIPTION
**Pull request by haim-kermany**
lots of changed files.
created two new kinds of tree nodes:
* GroupSquareTreeNode
* GroupPointTreeNode

Ziv, can you please review the drawio pkg code?
Adi, can you please review code at vpcmodel and ibmvpc?

### CODE REVIEW

File: pkg/drawio/layout.go [hunk 1]
```diff
@@ -96,13 +106,156 @@ func canShareCell(i1, i2 IconTreeNodeInterface) bool {
 	return true
 }
 
+func sortBySize(groups []SquareTreeNodeInterface) []SquareTreeNodeInterface {
+	sortedBySizeGroups := make([]SquareTreeNodeInterface, len(groups))
+	copy(sortedBySizeGroups, groups)
+	sort.Slice(sortedBySizeGroups, func(i, j int) bool {
+		return len(sortedBySizeGroups[i].(*GroupSquareTreeNode).groupies) > len(sortedBySizeGroups[j].(*GroupSquareTreeNode).groupies)
```
Comment (zivnevo) [inline pkg/drawio/layout.go:113]: This assumes all members of the input slice are of type GroupSquareTreeNode.
If not checking this explicitly, at least add a comment with this assumption.

File: pkg/drawio/squareTreeNode.go [hunk 2]
```diff
@@ -206,11 +200,53 @@ func NewSubnetTreeNode(parent *ZoneTreeNode, name, cidr, acl string) *SubnetTree
...
+	theSubnet groupSquareVisibility = iota
+	square
+	innerSquare
+	connectedPoint
+)
+
+type GroupSquareTreeNode struct {
+	abstractSquareTreeNode
+	groupies   []IconTreeNodeInterface
+	visibility groupSquareVisibility
```
Comment (zivnevo) [inline pkg/drawio/squareTreeNode.go:234]: Should we promote `visibility` to `abstractSquareTreeNode` and the respective getter/setter to `SquareTreeNodeInterface`? Will this help with subnet grouping?

**Review by zivnevo — CHANGES_REQUESTED** [review]
This is great code. Please see some comments

File: pkg/drawio/iconTreeNode.go [hunk 3]
```diff
@@ -170,6 +169,57 @@ func (tn *VsiTreeNode) DrawioParent() TreeNodeInterface {
 
 func (tn *VsiTreeNode) IsVSI() bool { return true }
 
+// ///////////////////////////////////////////
+type GroupPointTreeNode struct {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/iconTreeNode.go:178]: done

File: pkg/drawio/squareTreeNode.go [hunk 4]
```diff
@@ -206,11 +200,53 @@ func NewSubnetTreeNode(parent *ZoneTreeNode, name, cidr, acl string) *SubnetTree
...
+const (
+	theSubnet groupSquareVisibility = iota
+	square
+	innerSquare
+	connectedPoint
+)
+
+type GroupSquareTreeNode struct {
+	abstractSquareTreeNode
+	groupies   []IconTreeNodeInterface
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/squareTreeNode.go]: done

File: pkg/drawio/iconTreeNode.go [hunk 5]
```diff
@@ -170,6 +169,57 @@ func (tn *VsiTreeNode) DrawioParent() TreeNodeInterface {
...
+	groupPoint := &GroupPointTreeNode{
+		abstractIconTreeNode: newAbstractIconTreeNode(parent, ""),
+		directed:             directed,
+		isSrc:                isSrc,
+	}
+	parent.addIconTreeNode(groupPoint)
+	return groupPoint
+}
+func (tn *GroupPointTreeNode) connectGroupies() {
+	for _, groupe := range tn.Parent().(*GroupSquareTreeNode).groupies {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/iconTreeNode.go]: done

File: pkg/drawio/iconTreeNode.go [hunk 6]
```diff
@@ -170,6 +169,57 @@ func (tn *VsiTreeNode) DrawioParent() TreeNodeInterface {
...
+		var s, d IconTreeNodeInterface = tn, groupe
+		if tn.isSrc {
+			s, d = groupe, tn
+		}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/iconTreeNode.go]: done

File: pkg/drawio/layoutOverlap.go [hunk 7]
```diff
@@ -67,16 +69,55 @@ func (lyO *layoutOverlap) fixOverlapping() {
...
+			continue
+		}
+		col := src.Location().firstCol
+		lineX := col.x() + col.thickness/2 - linesOnCol[col]*widthBetweenLines
+		for _, gi := range []*GroupPointTreeNode{src, dst} {
+			if gi.hasShownSquare() {
+				// adding a point:
+				p := iconCenterPoint(gi)
+				line.addPoint(lineX, p.Y)
+			} else {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layoutOverlap.go:111]: I wanted to point out that both src and dst have the same X, and that it calculated once

File: pkg/drawio/layout.go (hunk 1 above)
Comment (haim-kermany) [reply pkg/drawio/layout.go]: done

File: pkg/drawio/layout.go [hunk 8]
```diff
@@ -96,13 +106,156 @@ func canShareCell(i1, i2 IconTreeNodeInterface) bool {
...
+			outerToInnersGroup[group] = map[SquareTreeNodeInterface]bool{}
+		}
+		for _, icon := range group.groupies {
+			if group.visibility == square {
+				iconOuterGroup[icon] = group
+			} else if group.visibility == innerSquare {
+				iconInnerGroup[icon] = group
+				outerToInnersGroup[iconOuterGroup[icon]][group] = true
+			}
+		}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layout.go]: done

File: pkg/drawio/squareTreeNode.go (hunk 2 above)
Comment (haim-kermany) [reply pkg/drawio/squareTreeNode.go]: didnt figure out the subnet grouping yet.
I might use there GroupSquareTreeNode too, so visibility should stay here

File: pkg/drawio/layoutOverlap.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/drawio/layoutOverlap.go:111]: makes sense

File: pkg/drawio/iconTreeNode.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply pkg/drawio/iconTreeNode.go]: got it

File: pkg/drawio/layout.go [hunk 9]
```diff
@@ -96,13 +106,158 @@ func canShareCell(i1, i2 IconTreeNodeInterface) bool {
 	return true
 }
 
+// sortGroupSquareBySize() sorting GroupSquareTreeNodes by the size of their groupedIcons.
+// assuming the input squares are sortedBySizeGroups
```
Comment (zivnevo) [inline pkg/drawio/layout.go:110]:
```suggestion
// sortGroupSquareBySize() sorts a slice of GroupSquareTreeNodes by the size of their groupedIcons set.
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/drawio/iconTreeNode.go [hunk 10]
```diff
@@ -170,6 +169,64 @@ func (tn *VsiTreeNode) DrawioParent() TreeNodeInterface {
...
+		isSrc:                isSrc,
+	}
+	parent.addIconTreeNode(groupPoint)
+	return groupPoint
+}
+func (tn *GroupPointTreeNode) connectGroupedIcons() {
+	for _, groupedIcon := range tn.Parent().(*GroupSquareTreeNode).groupedIcons {
+		var s, d IconTreeNodeInterface = tn, groupedIcon
+		// in case the GroupPoint is the src, its means we have an arrow from the tn to  its colleague.
+		// so we want to add errors fro the groupIcon the the GroupPoint
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/iconTreeNode.go]: done

**Review by zivnevo — DISMISSED** [review]

File: pkg/ibmvpc/analysis_output_test.go [hunk 11]
```diff
@@ -163,6 +163,32 @@ var tests = []*vpcGeneralTest{
 		format:   vpcmodel.Text,
 	},
 
+	// batch2.5: only vsi-level use-case, with grouping , drawio format
+	{
+		name:     "acl_testing3",
+		useCases: []vpcmodel.OutputUseCase{vpcmodel.AllEndpoints},
+		grouping: true,
+		format:   vpcmodel.DRAWIO,
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/ibmvpc/analysis_output_test.go:171]: done

**Review by adisos — APPROVED** [review]

**Pull request merged**